src/app/CommandHandlerDispatcher.cpp

```diff
@@ -0,0 +1,128 @@
+{
+    CHIP_ERROR err                           = CHIP_NO_ERROR;
+    bool sendStatusResponseWithInvalidAction = false;
+
+    if (mState == State::AwaitingResponse && aPayloadHeader.HasMessageType(Protocols::InteractionModel::MsgType::StatusResponse))
+    {
+        CHIP_ERROR statusError = CHIP_NO_ERROR;
+        err                    = StatusResponse::ProcessStatusResponse(std::move(aPayload), statusError);
+        VerifyOrExit(err == CHIP_NO_ERROR, sendStatusResponseWithInvalidAction = true);
+        statusError = err;
+        VerifyOrExit(err == CHIP_NO_ERROR, sendStatusResponseWithInvalidAction = true);
+
+        err = SendCommandResponse();
+        VerifyOrExit(err == CHIP_NO_ERROR, sendStatusResponseWithInvalidAction = true);
+
+        if (mState != State::AwaitingResponse)
+        {
+            // We are sending out the last message and no longer are expecting any responses. As a results we are
+            // fulfilling out responsibility to call close by calling ExitNow().
+            ExitNow();
+        }
+        return CHIP_NO_ERROR;
+    }
+    ChipLogDetail(DataManagement, "CommandHandler: Unexpected message type %d", aPayloadHeader.GetMessageType());
+    StatusResponse::Send(Status::InvalidAction, mExchangeCtx.Get(), false /*aExpectResponse*/);
+    err = CHIP_ERROR_INVALID_MESSAGE_TYPE;
+    if (mState == State::AwaitingResponse)
+    {
+        // We were waiting on a response and we got something we did not expect, so we will no longer continue sending
+        // messages, nor do we expect to get any more responses. As a result we are simply fulfilling our
+        // responsibility to call Close() by calling ExitNow().
+        ExitNow();
+    }
+    return err;
+exit:
+    if (sendStatusResponseWithInvalidAction)
+    {
+        StatusResponse::Send(Status::InvalidAction, mExchangeCtx.Get(), false /*aExpectResponse*/);
+    }
+    mpCommandHandler->Close();
+    return err;
+}
+
+void CommandHandlerDispatcher::OnResponseTimeout(Messaging::ExchangeContext * apExchangeContext)
+{
+    ChipLogDetail(DataManagement, "CommandHandlerDispatcher: Timed out waiting for response from requester mState=[%10.10s]",
+                  GetStateStr());
+    mpCommandHandler->Close();
+}
+
+CHIP_ERROR CommandHandlerDispatcher::SendCommandResponse()
+{
+    VerifyOrReturnError(!mChunks.IsNull(), CHIP_ERROR_INCORRECT_STATE);
+    System::PacketBufferHandle commandResponsePayload = mChunks.PopHead();
+
+    bool moreToSend               = !mChunks.IsNull();
+    Messaging::SendFlags sendFlag = Messaging::SendMessageFlags::kNone;
+    if (moreToSend)
+    {
+        sendFlag = Messaging::SendMessageFlags::kExpectResponse;
+    }
+
+    bool sendingFirstResponse = mState == State::Idle;
+
+    // If this is the first response we are sending, and we are waiting on response to send out another
+    // InvokeResponseMessages we setup the response timer.
```

Comment:
Ignoring the grammar/spelling errors in the comment: why worry about first time or not?  We can just set this whenever we are sending something expecting a response, no?

Reply:
I felt it was a little redundant to set it every time. But sure I will just set it every time